### ISSUE TITLE
[css-mixins] `<dashed-function>` syntax seems invalid

### ISSUE THREAD

**Issue by cdoublev** [w3c/csswg-drafts#10558]
> A [`<dashed-function>`](https://drafts.csswg.org/css-mixins-1/#typedef-dashed-function) is a functional notation whose function name starts with two dashes (U+002D HYPHEN-MINUS). Its syntax is:
  >
  > `--*() = --*( <declaration-value># )`

First, I think it should be `<dashed-function> = --*( <declaration-value># )`.

Second, the [CSS Value definition syntax](https://drafts.csswg.org/css-values-4/#value-defs) does not allow using `*` as a wildcard in a function's name.

Solutions I am thinking of:

  1. define the production rule as `informative`
  2. use syntax similar to [functional pseudo selector](https://drafts.csswg.org/selectors-4/#typedef-pseudo-class-selector): `<function-name> <declaration-value># ')'`

I prefer solution 1 because the syntax of solution 2 represents a list of tokens, whereas productions always represent a list of component values, because all entry points of the CSS parser consume component values from the input list of tokens: you do not match the input against `<function-name>`, then `<declaration-value>#`, then `<)-token>`, but against a function component value. Related: #7016.

**Comment by cdoublev** [comment]
Somewhat related... Sometimes specs:

  - disallow whitespace between component values, like `<`, `<type-name>`, and `>`
  - require whitespace, like between two `<complex-selector-unit>`s if the `<combinator>` between them is omitted

But it is not explicitly required after `using` in `using ( <dependency-list> )`. Obviously there would be a function if the whitespace was omitted, so this may not be worth clarifying. (Please tell me if I should open a separate issue.)

When a whitespace is missing between two symbols, eg. `<foo><bar>`, it does not mean that a whitespace in the input is disallowed. My point is that if whitespaces were defined to be optionals in a value definition if not explicitly disallowed...

  - after `<`
  - before `>`
  - before a multiplier
  - `(` preceding a symbol defining a function's name
  - `{` preceding a symbol defining a rule's name

... a syntax like `<function-name>(<declaration-value>#)` for `<dashed-function>` would be unambiguous.

Fwiw, I think this would be confusing and error-prone. A simpler solution is to rewrite with `using <dependency-block>`.

**Comment by tabatkins** [comment]
Re: your whitespace comment, there's a bit more, too. Here's the syntax, currently:

```
&lt;@function> = @function <<function-name>> [ ( <<parameter-list>> ) ]?
	[ using ( <<dependency-list>> ) ]?
	[ returns <<type>> ]?
{
	<<declaration-rule-list>>
}

In addition to the `using` detail (`using ()` is allowed, `using()` is illegal), the main parameters are *also* separate and optional, so `@function --foo ()` is allowed, but `@function --foo()` is illegal.

This seems odd; in JS, `function foo() {...}` is, I think, the most common syntactic form (even tho `function foo () {...}` is allowed in JS, too). I think we should probably at minimum switch to that; possibly we can allow both forms, if useful. This is a novel syntactic form, so even tho it's kinda unprecedented, that might still be acceptable.

@andruud , did you intend to write both of these locations with that whitespace, or was it unintended?

**Comment by andruud** [comment]
@cdoublev Thanks for reviewing this spec.

> `--*() = --*( <declaration-value># )`

Yeah, I had no idea how to express the thing that needed expressing here. This was inspired by [--*](https://drafts.csswg.org/css-variables/#defining-variables) from css-variables-1. `<dashed-function> = --*( <declaration-value># )` looks better.

What does it mean to mark it `informative`, and how do I do that?

> @andruud , did you intend to write both of these locations with that whitespace, or was it unintended?

Not exactly intentional. I based this on [Miriam's explainer](https://css.oddbird.net/sasslike/mixins-functions/#defining-a-function-the-function-rule), and then forgot to think about whitespace while copying that grammar.

I agree that `@function --foo()` should work. And _probably_ it should be the only valid form?

Not sure if that means we should require `using(` as well. Although really I'm hoping we can find an alternatively way to represent the dependencies which avoids the issue.

**Comment by cdoublev** [comment]
Tab already [marked it](https://github.com/w3c/csswg-drafts/commit/71d777bfa5f1ee2bbed3bf62ab2e776d64a1b820#diff-9c4c1c5a9da867a8f9b01470f5918ee9a736bce6ba707c484a34c6e51f8dffb4L216) as *non-normative* with the [`informative` class](https://speced.github.io/bikeshed/#metadata-informative-classes). Now `w3c/reffy` no longer extracts it from the spec.

If I am not mistaken, `<dashed-function>` will never be included in other value definitions, like other substitution functions. So this issue may therefore be considered fixed.

But since it must be syntactically validated, it would be nice to define its syntax using a *normative* value definition.

Fwiw, I like Tab's idea. This requires disallowing function names like `not`, `and`, `or`, right? This would fix #7016 by replacing `<function-token> <any-value> <)-token>` with `<ident> (<any-value>)`.

**Comment by andruud** [comment]
> So this issue may therefore be considered fixed.

@cdoublev Should we close it, then?

**Comment by cdoublev** [comment]
Yes, I think so. Since an implementation should parse a declaration value by looking for a function whose name is `var`, `attr`, `env`, etc... and now whose name starts with `--`, `<dashed-function>` does not need a valid value definition.

Other concerns can be addressed in a separate, more general issue. Namely:

  - how to specify grammar for a function with a value range (production) for its name? (related: #11241)
  - allow whitespaces between a function name and `(`

The first point is what bothers me a little. Since `<var()>` has a valid value definition, I can [parse something according to its grammar](https://drafts.csswg.org/css-syntax-3/#css-parse-something-according-to-a-css-grammar) with the string `<var()>` as `grammar`. I cannot do the same with `<dashed-function>`. It would be nice if this were possible.

**Comment by tabatkins** [comment]
> how to specify grammar for a function with a value range (production) for its name?

That's already done in the spec. You have to use some prose rather than a pure Value Definition Syntax, but that's fine.

> allow whitespaces between a function name and (

Nothing to be discussed here, no whitespace is allowed. That's consistent for all functions in CSS.

> I cannot do the same with `<dashed-function>`. It would be nice if this were possible.

Sure you can. `<dashed-function>` is defined in the Mixins spec, it's perfectly usable as a non-terminal.

**Comment by cdoublev** [comment]
> > allow whitespaces between a function name and (
  >
  > Nothing to be discussed here, no whitespace is allowed. That's consistent for all functions in CSS.

You started discussing this in the [comment](https://github.com/w3c/csswg-drafts/issues/10558#issuecomment-2226420917) above:

  > I think we should probably at minimum switch to that; possibly we can allow both forms, if useful. This is a novel syntactic form, so even tho it's kinda unprecedented, that might still be acceptable.

I do not want to continue a discussion that is already over, but if I see `fn (a)` in a value definition, is that allowed? does it represent a function or a keyword and a simple block?

  > Sure you can. `<dashed-function>` is defined in the Mixins spec, it's perfectly usable as a non-terminal.

It is true. I thought about the problem wrongly. Sorry.

**Comment by tabatkins** [comment]
> You started discussing this in the https://github.com/w3c/csswg-drafts/issues/10558#issuecomment-2226420917 above:

Yeah, note that my comment was that it was weird to allow `foo ()` and we probably shouldn't do it. ^_^

> if I see `fn (a)` in a value definition, is that allowed? does it represent a function or a keyword and a simple block?

That is an ident followed by a paranthesized block, no ambiguity.

**Comment by cdoublev** [comment]
> That is an ident followed by a paranthesized block, no ambiguity.

Also in `@layer <layer-name>? { <rule-list> }`? Does it represent `<at-keyword-token>`, `<layer-name>?`, and a simple block whose value is `<rule-list>`? And if I see `<production> (a)` in a value definition, is that an illegal value definition?

Similarly as whitespaces are (usually) optional in the input, I thought they should have no meaning in a value definition. `foo<bar>` or `<foo><bar>` are unambiguous, equivalent to `foo <bar>` or `<foo> <bar>`, and do not disallow a whitespace in the input. So that is not true for `fn(a)` and `fn (a)`.

I had not perceived that *[you] probably shouldn't do it"*, sorry. I am not arguing my personal opinion. At least, I just wanted to know yours, and ideally clarify it in the specification. It can even change in the future, that is ok.

**Comment by tabatkins** [comment]
> Does it represent `<at-keyword-token>`, `<layer-name>?`, and a simple block whose value is `<rule-list>`?

Yes. However, looking at the specs, I find there's not *actually* any definition of what an `@foo` means in a grammar definition. I guess it kind of grandfathered its way in without us realizing. I'll have to fix that.

> And if I see `<production> (a)` in a value definition, is that an illegal value definition?

No? That's perfectly valid, see <https://drafts.csswg.org/css-values/#component-types>. What makes you think it would be invalid?

> Similarly as whitespaces are (usually) optional in the input, I thought they should have no meaning in a value definition.

Note that while whitespace is usually optional, we still use greedy parsing semantics, so the CSS values `foo ()` and `foo()` are different things (ident followed by a paren block, and a function); you need to add whitespace (or comments) between many token pairs to keep them separate.  The same should be taken as true for the Value Definition Syntax.

However, it should be considered a spec authoring error if we ever rely on that; there should *always* be whitespace between different tokens in our grammars (with the exception that we sometimes don't add spaces around `[]` characters, if it reads better to have them wrapped "tightly", but they never produce ambiguous parses anyway).

**Comment by cdoublev** [comment]
> there's not actually any definition of what an `@foo` means in a grammar definition

I can find it in CSS V&U 4 (list item [1](https://drafts.csswg.org/css-values-4/#component-types)).

`# 123` is not equivalent to `#123`. `@ foo` is not equivalent to `@foo`. Etc. Good point. The syntax would be far less flexible if whitespaces were always meaningless to the parser.

`@foo {}` is equivalent to `@foo{}` but `foo ()` is not equivalent to `foo()`. A keyword and a simple block is consumed for both. From this point of view, it made sense to me to be consistent, especially now that function names are no longer always a simple identifier string to match case-insensitively, but also dashed identifiers thay may contain escaped characters to preserve in the serialization.

`<production> (a)` was a bad example.

**Comment by cdoublev** [comment]
> A keyword and a simple block is consumed for both.

Never mind, this is not true: a `<funcion-token>` is consumed for `fn()`. Ok.

**Comment by tabatkins** [comment]
Right, `foo(` is a distinct token, which is why it acts differently than the others. `foo()` is a function-token followed by a right parenthesis delim token.

(Tho I note you mentioned escapes as apparently being something new, but they've always been part of functions - you can write `\6c inear-gradient(90deg, red, blue)`. The ident part of a function token just uses normal ident-parsing rules.)

**Comment by cdoublev** [comment]
> you mentioned escapes as apparently being something new, but they've always been part of functions - you can write `\6c inear-gradient(90deg, red, blue)`. The ident part of a function token just uses normal ident-parsing rules.

I am not sure I understand what you were trying to convey to me here. But from re-reading my own comment, I should have expanded on my point a little.

When parsing a function against `<dashed-function>`, an implementation simply needs to look for a function whose name starts with `--`. It is already guaranteed to be a valid `<ident>`. It matches `<dashed-ident>`. Assuming more function names were defined to match some production, implementations may choose to internally represent it with the result from parsing the name against this production. This would allow them to later apply serialization rules specific to this production, like re-escaping code points for `<dashed-ident>` and `<custom-ident>`.

`--cus\\ tom()` is tokenized then consumed as a function whose name is the string `--cus tom`, so implementation must either serialize it as an identifier, or they should escape it in their internal representation.
